### ISSUE TITLE
filedescriptors overrun

### ISSUE THREAD

**Issue by xstaticxgpx** [pexpect/pexpect#221]
Hi Guys,

This is partially related to #47

I've been using pexpect, wrapped inside a threaded python application, to help automate management of several thousand remote devices. It's been working great, but recently I'm starting to notice an issue where the filedescriptors opened by pexpect are seemingly never closed, or at least not closed promptly. This is causing me to run into the 1024 FD limit exception from select()

Doing a simple list of /proc/\<pid\>/fd will show the descriptors growing quickly over time, and eventually hitting 1024 entries which causes my application to start failing.

I'm fairly certain this is caused by me, badly integrating pexpect into my application, right now I have it so all the work is done in the pexpect session, then it tries to do a soft logout (sendline "exit" etc), then checks if the child is still alive and executes a terminate() - is there a better way to ensure the sessions are closed?

Perhaps this is a legitimate bug? My python application is only set to spawn 64 threads at the moment, it can't be possible that more than 64 pexpect sessions are running at any given time yet the file descriptors are going crazy. If it's not a bug I must be failing to end the sessions properly somehow.

Thanks for any help!

**Comment by xstaticxgpx** [comment]
I might have figured this out. The child.terminate() method I was using doesn't appear to work as I had expected,

I replaced it by referencing the __del__() method from spawn() class directly, and now the FDs are hovering at ~85-90 without growing beyond that at all. Same 64 thread count. I've confirmed by reverting this change that this indeed seems to have had a very noticeable effect.

```python
try:
    ap.child.__del__()
except:
    pass
```

as opposed to:
```python
try:
    if ap.child.isalive():
        ap.child.terminate(force=True)
except:
    pass
```

Another issue I had because of the overrun filedescriptors was that my application would "complete" (log messages showed it was complete) - but it would not actually terminate for several seconds. I've seen that issue before, but now that I've become aware of the file descriptor count issue, I saw that file descriptors were still being cleared from /proc/\<pid\>/fd during hangs like that - once the FDs got low enough the application would close. When there were the maximum 1024 file descriptors it would take maybe 15, 30 seconds to clear out and then finally my application would terminate. Ridiculous.
